packages/superset-ui-control-utils/src/types.ts

```diff
@@ -164,24 +159,33 @@ export type TabOverride = 'data' | boolean;
  *    be visibile.
  */
 export interface GeneralControlConfig {
-  type: InternalControlType | React.ComponentType;
+  type: unknown;
```

Comment:
Why is this changed to `unknown`?

Reply:
It is expanded later in more specific types. `GeneralControlConfig` is basically the base interface.

Reply:
Then do we need this field? Perhaps can just remove if this is only the base interface?

Reply:
Also, this is broader question. I often seen `type` field being string literal or something used as differentiator of union type. 
* Curious how is this `type` field used? 
* What is `internalControlType`? Is it a string?

Reply:
`InternalControlType` is a union of strings, basically the name of all the controls seeing in [controls/index.js](https://github.com/apache/incubator-superset/blob/291306392443a5a0d0e2ee0cc4a95d37c56d4589/superset-frontend/src/explore/components/controls/index.js#L73).

It is used make sure users are using known control types when passing string literals.

Reply:
I'm creating separate types instead of unions in the general type because I want TypeScript be able to infer requirements for other props based on what's being specified for `type`.

Also wanted to make sure `type` a required field for `GeneralControlConfig` so I think I'm going to keep it...

Reply:
I don't mean making `type` a union type. Let me clarify. Usually I do something like this. (also not sure if your "base interface" and my "base interface" mean the same thing)

```ts
interface BaseHero {
  attack: number;
  defense: number;
  speed: number;
  HP: number
} 

export interface Barbarian extends BaseHero {
  type: 'barbarian';
  fury: number;
}

export interface Sorceress extends BaseHero {
  type: 'sorceress';
  mana: number;
}

export type Hero = Barbarian | Sorceress;
```

Reply:
Yeah, the idea is the same. Maybe I can rename `GeneralControlConfig` to `BaseControlConfig` so people would no not to import it directly, then it feels more OK to remove `type`.

Reply:
Do we need to `export` the `BaseControlConfig`?

Reply:
Probably not

Reply:
Actually `type: unknown` is also needed, otherwise `BaseControlConfig & { type: xxx }` will report type incompatible error.

I made `BaseControlConfig` a generic type instead.